### PR TITLE
Trying to update to dukpy v0.5.0 with fixing the osx issue

### LINKED ISSUE #58: Dukpy for 3.12
Closed by: https://github.com/conda-forge/dukpy-feedstock/pull/16

Fixes: https://github.com/conda-forge/pyload-ng-feedstock/pull/57
Fixes: #69

### PR DESCRIPTION
**Pull request by HeavyTony2**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

Closes: #19 
Basically the same recipe than in PR 19, but with setuptools for the building error with osx version
Fixes:  https://github.com/conda-forge/pyload-ng-feedstock/issues/58

### CODE REVIEW

**Comment by HeavyTony2** [comment]
@conda-forge-admin, please rerender

**Comment by HeavyTony2** [comment]
Nevermind, I underesimated the issue, this needs a patch of this file for the osx_arm64 https://github.com/amol-/dukpy/blob/master/src/duktape/duk_config.h
```
 In file included from ./src/duktape/duktape.h:205:
  ./src/duktape/duk_config.h:459:7: error: 'TARGET_OS_BRIDGE' is not defined, evaluates to 0 [-Werror,-Wundef-prefix=TARGET_OS_]
```
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1079871&view=logs&j=a7e3c806-6468-5971-7aaf-ed365e039222&t=3200f604-54ec-52b8-7dd5-9c70be972e49&l=1280

**Comment by HeavyTony2** [comment]
@conda-forge-admin, please rerender

**Comment by HeavyTony2** [comment]
@pvanheus I managed to find what was the problem: the default clang version

**Comment by HeavyTony2** [comment]
@conda-forge/dukpy-feedstock

**Comment by HeavyTony2** [comment]
@conda-forge-admin, please rerender

**Comment by HeavyTony2** [comment]
@conda-forge-admin, please rerender

**Comment by HeavyTony2** [comment]
> @pvanheus I managed to find what was the problem: the default clang version

@conda-forge-admin, please ping team

**Comment by HeavyTony2** [comment]
Ping  @conda-forge/dukpy

**Comment by HeavyTony2** [comment]
Ping @conda-forge/dukpy

**Comment by HeavyTony2** [comment]
@conda-forge-admin, please rerender

**Comment by HeavyTony2** [comment]
https://conda-forge.zulipchat.com/#narrow/channel/457337-general/topic/PR.20not.20approved.20by.20the.20maintainer.20despite.20pings.20for.20a.20month

**Comment by HeavyTony2** [comment]
@conda-forge-admin, please rerender

**Comment by HeavyTony2** [comment]
@conda-forge/core Can you merge it, please?

File: recipe/conda_build_config.yaml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+c_compiler_version:
+  - 15  # [arm64]
```
Comment (hmaarrfk) [inline recipe/conda_build_config.yaml]:
```suggestion
  - 15                                 # [osx and arm64]
```

Comment (hmaarrfk) [reply recipe/conda_build_config.yaml]: can you add a comment as to what you saw without this? its somewhat strange that you got a terrible error without this.

Comment (HeavyTony2) [reply recipe/conda_build_config.yaml]: @hmaarrfk I think I'd have to temporarly remove the conda_build_config.yaml. I can't remember exactly what was the error message and the test logs got wiped away since it was more than one months ago. It could be related to 
```
 In file included from ./src/duktape/duktape.h:205:
  ./src/duktape/duk_config.h:459:7: error: 'TARGET_OS_BRIDGE' is not defined, evaluates to 0 [-Werror,-Wundef-prefix=TARGET_OS_]
```
but I'm not totally sure. In fact, perhaps this was an error related related to the Maxosx-SDK and not Clang itself. I've some flashback while browsing my search history. So perhaps the problem wasn't directly clang itself.

Comment (HeavyTony2) [reply recipe/conda_build_config.yaml]: We're gonna see what was the error message https://github.com/conda-forge/dukpy-feedstock/pull/20/commits/555a4a3256292d4473d9f0e5a6c3e52fdfb47e1b

**Comment by hmaarrfk** [comment]
@conda-forge-admin please rerender

**Comment by hmaarrfk** [comment]
(you need to rerender after "supressing")

**Comment by HeavyTony2** [comment]
> @conda-forge-admin please rerender

Thanks, I was puzzled about why it was suddenly working, then I remember i forgot to rerender, I was updating my local conda-smithy for a local rerendering 😅

**Comment by HeavyTony2** [comment]
@conda-forge-admin please rerender

File: recipe/conda_build_config.yaml (hunk 1 above)
Comment (HeavyTony2) [reply recipe/conda_build_config.yaml]: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1117240&view=logs&j=7df89df7-19f1-5ccc-b210-32c1d05c0f56&t=c5e4c004-2534-57b9-b739-c5013de31afa&l=1288

**Comment by HeavyTony2** [comment]
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1117240&view=logs&j=7df89df7-19f1-5ccc-b210-32c1d05c0f56&t=c5e4c004-2534-57b9-b739-c5013de31afa&l=1288

**Comment by hmaarrfk** [comment]
I fear that unfortunately don’t understand the OSX arm failure enough to comment if this is an appropriate fix

**Comment by HeavyTony2** [comment]
@conda-forge-admin please rerender

**Comment by hmaarrfk** [comment]
https://github.com/amol-/dukpy/blob/8de6f5072512e85f618ba383f5f68ae5b049a1f7/src/duktape/duk_config.h#L445


Maybe just patch this out.

**Comment by HeavyTony2** [comment]
> I fear that unfortunately don’t understand the OSX arm failure enough to comment if this is an appropriate fix

For what I understand, the "TARGET_OS_BRIDGE" thing is defined by TargetConditionals.h, but I can't tell which package on Conda contain this file.

**Comment by hmaarrfk** [comment]
Hmmm. This poses a clue!!!!

https://github.com/htacg/tidy-html5/issues/1030

**Comment by HeavyTony2** [comment]
```
[Clang Frontend Potentially Breaking Changes](https://releases.llvm.org/18.1.6/tools/clang/docs/ReleaseNotes.html#id208)
Target OS macros extension A new Clang extension (see [here](https://releases.llvm.org/18.1.6/tools/clang/docs/ReleaseNotes.html#target-os-detail)) is enabled for Darwin (Apple platform) targets. Clang now defines TARGET_OS_* macros for these targets, which could break existing code bases with improper checks for the TARGET_OS_ macros. For example, existing checks might fail to include the TargetConditionals.h header from Apple SDKs and therefore leaving the macros undefined and guarded code unexercised.

Affected code should be checked to see if it’s still intended for the specific target and fixed accordingly.

The extension can be turned off by the option -fno-define-target-os-macros as a workaround.
```
Ok, the thing with TargetConditionals.h is definetely about Clang 18. https://releases.llvm.org/18.1.6/tools/clang/docs/ReleaseNotes.html#clang-frontend-potentially-breaking-changes

**Comment by HeavyTony2** [comment]
@conda-forge-admin please rerender

**Comment by HeavyTony2** [comment]
@hmaarrfk So the issue is caused by a change with Clang 18.1.6. It's working with Clang 18.1.5. What is considered as the more appropriate? Pinning the 18.1.5 version in conda_build_config.yaml or patching files?

**Comment by hmaarrfk** [comment]
Maybe just pin to what you need and open an issue on the clang feedstock.

**Comment by HeavyTony2** [comment]
Is this a good explanatory comment? https://github.com/conda-forge/dukpy-feedstock/blob/81cd10dab0e9ea7c26a4cd86450a3f4f67911607/recipe/conda_build_config.yaml#L3

**Comment by hmaarrfk** [comment]
The explanation is fine. But you need to open an issue somewhere to ensure it is sustainable. Maybe on the dukpy repo???

File: recipe/conda_build_config.yaml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+c_compiler_version:
+  - 18.1.5                                 # [osx and arm64]
```
Comment (h-vetinari) [inline recipe/conda_build_config.yaml]: 18.1.5 is not receiving any updates in our infrastructure. If you need to stay below 18.1.6, better to just use clang 17.

However, the clang release notes you found (well done!) point out a much better solution:
> The extension can be turned off by the option `-fno-define-target-os-macros` as a workaround.

Comment (h-vetinari) [reply recipe/conda_build_config.yaml]: So you should be able to do something like
```
export CFLAGS="$CFLAGS -fno-define-target-os-macros"
```
in the build script

Comment (HeavyTony2) [reply recipe/conda_build_config.yaml]: How can I proceed with the build script? Should I make a build.sh specifically for osx?

Comment (HeavyTony2) [reply recipe/conda_build_config.yaml]: Nevermind, I took a look into some other recipe in conda-forge and I can add it to the meta.yaml

**Comment by HeavyTony2** [comment]
@conda-forge-admin please rerender

**Comment by HeavyTony2** [comment]
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1117375&view=logs&j=a7e3c806-6468-5971-7aaf-ed365e039222&t=3200f604-54ec-52b8-7dd5-9c70be972e49&l=1069
`- export CFLAGS="${CFLAGS} -fno-define-target-os-macros"  # [osx and arm64]` Seems to be fruitless

**Comment by h-vetinari** [comment]
> `- export CFLAGS="${CFLAGS} -fno-define-target-os-macros" # [osx and arm64]` Seems to be fruitless

You're looking at the wrong point in the logs. What you linked is the default value upon activation of the compilers -- conda just shows you all the environment variables that are set before starting the build. Whatever modifications you make necessarily come after that.

In the actual stacktrace you can see that the flag is present though. That it doesn't work could be due to any number of reasons, but I don't have time to chase down this rabbit hole; all told, going back to clang v17 and raising an upstream issue is probably the best route.

**Comment by HeavyTony2** [comment]
> going back to clang v17 and raising an upstream issue is probably the best route.

So, do I raise the issue on dukpy repo or on llvm/clang/clangdev feedstock?

**Comment by HeavyTony2** [comment]
Situation addressed in Dukpy repo https://github.com/amol-/dukpy/issues/82
I also let comment in the conda_build_config.yaml indicating this issue is adressed upstream
https://github.com/conda-forge/dukpy-feedstock/blob/bd6078bb66b8a2a05edb596fe6ed749d042efdc0/recipe/conda_build_config.yaml#L3-L4

**Comment by HeavyTony2** [comment]
Fun fact
https://github.com/amol-/dukpy/blob/8de6f5072512e85f618ba383f5f68ae5b049a1f7/src/duktape/duk_config.h#L459
Isn't there in the last built version on conda
https://github.com/amol-/dukpy/blob/bde1d080871936f2cec6f7e830c25cddc0ba2bee/src/duktape/duk_config.h#L459

Dukpy 0.2.3 is buildable with clang 18 as confirmed by https://github.com/conda-forge/dukpy-feedstock/pull/21

Which means the issue is definetely with the Dukpy code.

**Comment by hmaarrfk** [comment]
good enough for me for you to take over. 

h-vetinari might have other comments, but you can probably request help in a separate cleanup PR.

Thank you for raising the issue upstream.
